resources/testdriver.js

```diff
@@ -137,27 +137,64 @@
         }
     };
 
-    window.test_driver_internal = {
+    var manual = {
         /**
-         * Triggers a user-initiated click
+         * This flag should be set to `true` by any code which implements the
+         * internal methods defined below for automation purposes. Doing so
+         * allows the library to signal failure immediately when an automated
+         * implementation of one of the methods is not available.
+         */
+        in_automation: false,
+
+        /**
+         * Waits for a user-initiated click
          *
          * @param {Element} element - element to be clicked
          * @param {{x: number, y: number} coords - viewport coordinates to click at
-         * @returns {Promise} fulfilled after click occurs or rejected if click fails
+         * @returns {Promise} fulfilled after click occurs
          */
         click: function(element, coords) {
-            return Promise.reject(new Error("unimplemented"));
+            if (window.test_driver_internal.in_automation) {
+                return Promise.reject(new Error('Not implemented'));
+            }
+
+            return new Promise(function(resolve, reject) {
+                element.addEventListener("click", resolve);
+            });
         },
 
         /**
-         * Triggers a user-initiated click
+         * Waits for an element to receive a series of key presses
          *
-         * @param {Element} element - element to be clicked
-         * @param {String} keys - keys to send to the element
-         * @returns {Promise} fulfilled after keys are sent or rejected if click fails
+         * @param {Element} element - element which should receve key presses
+         * @param {String} keys - keys to expect
+         * @returns {Promise} fulfilled after keys are received or rejected if
+         *                    an incorrect key sequence is received
          */
         send_keys: function(element, keys) {
-            return Promise.reject(new Error("unimplemented"));
+            if (!window.test_driver_internal.in_automation) {
+                return Promise.reject(new Error('Not implemented'));
+            }
+
+            return new Promise(function(resolve, reject) {
+                var seen = "";
+
+                function onKeyDown(event) {
+                    if (event.key.length > 1) {
+                        return;
+                    }
+
+                    seen += event.key;
+
+                    if (keys.indexOf(seen) !== 0) {
+                        reject(new Error("Unexpected key sequence: " + seen));
+                    } else if (seen === keys) {
+                        resolve();
+                    }
+                }
+
+                element.addEventListener("keydown", onKeyDown);
+            });
```

Comment:
Can you add a `.finally` that removes the event listener? Otherwise that `seen` variable will keep building up for every keydown, which is harmless, but takes effort for the reader to understand that it is.

Reply:
`finally` is as old as exceptions in ES are (i.e., ES3), and I don't think anything ever supported exceptions but not a `finally` clause?

Reply:
The `finally` that @foolip and I are discussing is `Promise.prototype.finally` as standardized in ES2018